### PR TITLE
Cross-version API Compatibility tests - Config Metadata APIs

### PR DESCRIPTION
**Pull request by mpanchajanya**
### What this PR does / why we need it

- Added Config Metadata APIs compatibility tests  for SetCLIDiscoverySource, GetCLIDiscoverySource, DeleteCLIDiscoverySource APIs for Runtime libraries latest, v0.28.0
- Added helper methods to create API Commands, Validators to validate input and outputOptions.
- Extended the testplugins implementation to trigger  Config Metadata APIs.
- Added Default helper functions to create the input and output options data for command creation

Detailed Info:

`test/compatibility`
`core/api_constants.go` -> Added  Config Metadata APIs names

`framework` changes
`compatibility-tests/metadata` -> Add compatibility tests for  Config Metadata APIs

`metadata` directory to provide functions to create api commands and validations to help write compatibility tests fo  Config Metadata APIs

`testplugins` changes
Add Implementation to trigger  Config Metadata APIs for all supported runtime versions

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

- Added Unit testing for the helper functions like Command creations, Trigger  Config Metadata APIs.
- Build and Run compatibility tests using below command and make sure all tests pass
`make compatibility-tests`

```
Ginkgo ran 6 suites in 25.350487917s
Test Suite Passed


```

### Release note

```release-note
Added config metadata APIs compatibility tests
```

### CODE REVIEW

File: test/compatibility/framework/metadata/metadata_commands.go [hunk 1]
```diff
@@ -0,0 +1,595 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package metadata provides api command helpers and validators to write compatibility tests for metadata apis
+package metadata
+
+import (
+	"strconv"
+
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/test/compatibility/core"
```
Comment (vuil) [inline test/compatibility/framework/metadata/metadata_commands.go:10]: nit: L10 and L12 order should be flipped

File: test/compatibility/framework/metadata/metadata_commands.go [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+	"gopkg.in/yaml.v3"
+)
+
+// NewSetConfigMetadataPatchStrategyCommand constructs a command to make a call to specific runtime version SetConfigMetadata API
+// Input Parameter inputOpts has all input parameters which are required for Runtime SetConfigMetadata API
```
Comment (vuil) [inline test/compatibility/framework/metadata/metadata_commands.go:16]: nit: missing colon to be consistent with the rest of the comment?

File: test/compatibility/framework/metadata/metadata_commands.go [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+// Input Parameter: outputOpts has details about expected output from Runtime SetConfigMetadata API call
+// Return: command to execute or error if any validations fails for SetConfigMetadataInputOptions or SetConfigMetadataOutputOptions
+// This method does validate the input parameters  SetConfigMetadataInputOptions or SetConfigMetadataOutputOptions based on Runtime API Version
```
Comment (vuil) [inline test/compatibility/framework/metadata/metadata_commands.go:19]: nit: double space

File: test/compatibility/framework/metadata/metadata_commands.go [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+// For more details about supported parameters refer to SetConfigMetadataInputOptions or SetConfigMetadataOutputOptions definition(and ConfigMetadataOpts struct, which is embedded)
```
Comment (vuil) [inline test/compatibility/framework/metadata/metadata_commands.go:20]: space before (

This, and the nits above apply several other functions below as well.

File: test/compatibility/framework/metadata/metadata_helpers.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+// Metadata related constants
+
+const (
+	CompatibilityTestsMetadataPatchStrategyKey   = "compatibility-tests.contexts.name"
+	CompatibilityTestsMetadataPatchStrategyValue = "replace"
+	CompatibilityTestsMetadataSettingsKey        = "useUnifiedConfig"
+	CompatibilityTestsMetadataSettingsValue      = "true"
+)
+
+// DefaultSetConfigMetadataPatchStrategyCommand creates a GetEnvConfigurations Command with default input and output options
```
Comment (vuil) [inline test/compatibility/framework/metadata/metadata_helpers.go:24]: typo: GetEnvConfigurations, same for several functions below

File: test/compatibility/testplugins/runtime-test-plugin-latest/go.work.sum [hunk 6]
```diff
@@ -5,8 +5,11 @@ github.com/go-logr/logr v1.2.3/go.mod h1:jdQByPbusPIv2/zmleS9BjJVeZ6kBagPoEUsqbV
 github.com/golang/protobuf v1.5.2/go.mod h1:XVQd3VNwM+JqD3oG2Ue2ip4fOMUkwXdXDdiuN0vRsmY=
 github.com/google/pprof v0.0.0-20210407192527-94a9f03dee38/go.mod h1:kpwsk12EmLew5upagYY7GY0pfYCcupk39gWOCRROcvE=
 github.com/ianlancetaylor/demangle v0.0.0-20200824232613-28f6c0f3b639/go.mod h1:aSSvb/t6k1mPoxDqO4vJh6VOCGPwU4O0C2/Eqndh1Sc=
+github.com/inconshreveable/mousetrap v1.1.0/go.mod h1:vpF70FUmC8bwa3OWnCshd2FqLfsEA9PFc4w1p2J65bw=
```
Comment (vuil) [inline test/compatibility/testplugins/runtime-test-plugin-latest/go.work.sum:8]: Surprised this is only needed for -latest

File: test/compatibility/framework/metadata/metadata_commands.go (hunk 1 above)
Comment (mpanchajanya) [reply test/compatibility/framework/metadata/metadata_commands.go:10]: updated

File: test/compatibility/framework/metadata/metadata_commands.go (hunk 2 above)
Comment (mpanchajanya) [reply test/compatibility/framework/metadata/metadata_commands.go:16]: updated

File: test/compatibility/framework/metadata/metadata_commands.go (hunk 3 above)
Comment (mpanchajanya) [reply test/compatibility/framework/metadata/metadata_commands.go:19]: removed

File: test/compatibility/testplugins/runtime-test-plugin-latest/go.work.sum (hunk 6 above)
Comment (mpanchajanya) [reply test/compatibility/testplugins/runtime-test-plugin-latest/go.work.sum:8]: updated all go mods

**Review by vuil — APPROVED** [review]
lgtm, thanks for the updates

**Pull request merged**